### ISSUE TITLE
isinstance inference not working in closures

### ISSUE THREAD

**Issue by ThomasHickman** [python/mypy#4679]
In the code:
```python
class SuperCls:
    pass

class Cls(SuperCls):
    prop = 9

def do_something(inp):
    # type: (SuperCls) -> None
    if isinstance(inp, Cls):
        def x():
            # type: () -> None
            print(inp.prop)
        return x()

inst = Cls()
do_something(inst)
```
this warning is wrongly emitted:
```
test.py:12: error: "SuperCls" has no attribute "prop"
```

**Comment by gvanrossum** [comment]
This is a known issue; it's somewhat of a feature because sometimes the closure variable can change its value (and hence type) again before the function is called (e.g. consider if your code has `inp = SuperCls()` just before the `return x()`). Detecting the cases where this isn't a problem is somewhat complicated. You can work around it by passing `inp` as an argument to `x()`.

**Comment by nilehmann** [comment]
Related to this. I have code similar to this one

```
def add(m: int, n: int) -> int:
    return m + n

one: Optional[int] = 1

if one:
    map(lambda n: add(one, n), [1, 2, 3])
```
Maybe the case for anonymous function is easier?

**Comment by nilehmann** [comment]
Giving it a second thought, I think is it equally easy to invaidates things for lambdas than for regular functions.

**Comment by jrideout** [comment]
I ran into this as well:

```
def test(a: Optional[str] = None) -> str:
    if a is None:
        a = 'abc'
    def inner() -> str:
        return a  #  <---  error: Incompatible return value type (got "Optional[str]", expected "str")
    return inner()
```

In this cases the closure isn't going to change. I wonder if we can infer that.

**Comment by ilevkivskyi** [comment]
This is actually a duplicate of https://github.com/python/mypy/issues/2608 (already high priority). Yes, one of proposed solutions was to use the binder, if e.g. the inner function appears at the end of the outer function, so the closure can't change.
